tools/cpudist.py

```diff
@@ -52,60 +55,89 @@
 bpf_text += tp.generate_entry_probe()
 bpf_text += tp.generate_struct()
 
+if not args.offcpu:
+    bpf_text += "#define ONCPU\n"
+
 bpf_text += """
 typedef struct pid_key {
     u64 id;
     u64 slot;
 } pid_key_t;
 
-// We need to store the start time, which is when the thread got switched in,
-// and the tgid for the pid because the sched_switch tracepoint doesn't provide
-// that information.
+
 BPF_HASH(start, u32, u64);
 BPF_HASH(tgid_for_pid, u32, u32);
 STORAGE
 
+static inline u32 get_tgid(u32 pid)
+{
+    u32 *stored_tgid = tgid_for_pid.lookup(&pid);
+    if (stored_tgid != 0)
+        return *stored_tgid;
+    return 0xffffffff;
+}
+
+static inline void store_start(u32 tgid, u32 pid, u64 ts)
+{
+    if (FILTER)
+        return;
+
+    start.update(&pid, &ts);
+}
+
+static inline void update_hist(u32 tgid, u32 pid, u64 ts)
+{
+    if (FILTER)
+        return;
+
+    u64 *tsp = start.lookup(&pid);
+    if (tsp == 0)
+        return;
+
+    u64 delta = ts - *tsp;
+    FACTOR
+    STORE
+}
+
 int sched_switch(struct pt_regs *ctx)
 {
+    u64 ts = bpf_ktime_get_ns();
     u64 pid_tgid = bpf_get_current_pid_tgid();
+    u32 tgid = pid_tgid >> 32, pid = pid_tgid;
+    // Keep a mapping of tgid for pid because when sched_switch hits,
+    // we only have the tgid information for the *current* pid, but not
+    // for the previous one.
+    tgid_for_pid.update(&pid, &tgid);
+
     u64 *di = __trace_di.lookup(&pid_tgid);
     if (di == 0)
         return 0;
 
     struct sched_switch_trace_entry args = {};
     bpf_probe_read(&args, sizeof(args), (void *)*di);
 
-    u32 tgid, pid;
-    u64 ts = bpf_ktime_get_ns();
-
+#ifdef ONCPU
     if (args.prev_state == TASK_RUNNING) {
-        pid = args.prev_pid;
-
-        u32 *stored_tgid = tgid_for_pid.lookup(&pid);
-        if (stored_tgid == 0)
-            goto BAIL;
-        tgid = *stored_tgid;
-
-        if (FILTER)
+#else
+    if (1) {
+#endif
+        u32 prev_pid = args.prev_pid;
+        u32 prev_tgid = get_tgid(prev_pid);
```

Comment:
actually comparing to TASK_RUNNING is probably doing not what you expected. Take a look at __trace_sched_switch_state(). It will set tracepoint's pref_state to TASK_RUNNING | TASK_STATE_MAX and our 'if' will miss it. Looking at prev->state is cleaner.